packages/browser-data-collector/src/should-send.ts

```diff
@@ -21,5 +21,5 @@
  */
 // eslint-disable-next-line @typescript-eslint/no-unused-vars
 export const shouldSend = ( _report: Report ): boolean => {
-	return true;
+	return Math.random() <= 0.05;
```

Comment:
The PR description says:

> Enable RUM tracking in production to 5% users

but that's slightly misleading 🙂 This code selects random 5% from all the individual reports and sends only them.

"5% users" would mean sending all reports for 5% of user accounts and zero reports for everyone else. We have stats and A/B tests that behave this way.

Also, hard-coding the condition into the `browser-data-collector` is something I'd frown upon. Something that claims to be an independent package should have higher standards of reusability and flexibility.

Reply:
> We have stats and A/B tests that behave this way.

How does that work? Maybe I can reuse that system.

> hard-coding the condition into the browser-data-collector

Agreed. I'll think how we can change the API so this can be specified in a clean and nice way from Calypso side.

Reply:
Check out the abtest lib here:
https://github.com/Automattic/wp-calypso/tree/master/client/lib/abtest

Reply:
> How does that work? Maybe I can reuse that system.

The code that assigns A/B tests variations is here:

https://github.com/Automattic/wp-calypso/blob/master/client/lib/abtest/index.js#L310-L335

There are two methods employed that make the assignments more stable:
- use the user ID instead of a random number. If 5% users are eligible, then it's users with (numerical) IDs that end with `*00`--`*04` and these with IDs that end with `*05`--`*99` are out. If we wanted the assignment to be less predicable, we could seed a random number generator with the user ID
- do the assignment only once and the cache it to local storage, or as a property in user's settings. This cached info can sometimes be lost or overwritten.

To store that A/B test assignments to the server, there is a `/me/abtests` REST API endpoint that you can GET from or POST to. The [WP.com API console](https://developer.wordpress.com/docs/api/console/) is a neat tool to play with that.

Reply:
Very cool library, I really like we can force ourselves in an experiment, it will help with testing.

It will limit our options to have different sampling rate based on the report properties though (eg: collect 10% of chrome, but 50% of firefox so we get statistically significant numbers for both). But we'll deal with that problem if we actually need it.

Reply:
> It will limit our options to have different sampling rate based on the report properties though (eg: collect 10% of chrome, but 50% of firefox so we get statistically significant numbers for both).

That would mean that we don't really need to assign a certain user to one group and keep them there. Throwing a dice on every individual report is also OK if that's what we want. Then we can have different quotas for different user agents.

Reply:
Looks like the lib doesn't support sending 5% of reports, 5% of page loads or even 5% of users, it always will be 5% of _browsers_.

> By default, test variations are assigned using a random number generator

If I'm reading the code correctly, this means that by default it will generate a random number, assign a variation and store it in `localStorage` forever, so all future visits with that browser profile (even if they are from a different WPCOM user) will get the same variation.

Probably not important for now, but worth remembering.

Reply:
> all future visits with that browser profile (even if they are from a different WPCOM user) will get the same variation.

If that's indeed true, I'm afraid it's a bug that should be fixed. We should include the user ID as part of the local storage keys, so that the current Calypso session doesn't load data stored by another user. Persisting Redux state to IndexedDB already works that way.

Some things that I'm not sure about:
- when the user is logged in, the A/B test data are part of the bootstrapped user info that's send by the server in _index.html_. Maybe that takes precedence over the local storage data, making local storage relevant only for logged-out sessions?
- if A/B test variations are assigned during a logged-out sessions, are they later "adopted" when the user logs in?

Both points are super-relevant during signups, where the session starts as logged-out and becomes logged-in at some point. Also, signup is the place where do a _lot_ of A/B tests and data collection.